packages/reactor/docs/planning/Operations/upgrades.md

```diff
@@ -0,0 +1,76 @@
+# Upgrades
+
+### Summary
+
+Document Model upgrades provide a systematic approach to versioning and backward compatibility.
+
+In Reactor, Actions act much like typical Command objects in the Event Sourcing architecture and Operations act much like Events.
+
+However, unlike typical ES approaches, Operations _are never upcasted_. That is, version upgrades never mutate Operations. Instead, every `Action` must be executed by a specific version of the corresponding document model library to result in the expected `Operation`.
+
+Operations are then decoupled from versioning, as the `Action` holds this information.
+
+### Document Creation Flow
+
+Suppose we have a Document Model of type `ph/todo`.
+
+1. First, we create a new document:
+
+```tsx
+import { createDocument } from "ph/todo";
+
+const doc = await createDocument({
+  title: "My Todo List",
+});
+```
+
+2. Next, we submit it to the `IReactor`.
+
+```tsx
+
+const jobStatus = await reactor.create(doc);
+
+```
+
+3. Internally, the `IReactor` will:
+
+- Create an `Action` of type `ACTION_CREATE`.
+- Create an `Action` of type `ACTION_UPGRADE`.
+- Submit both actions in a single `Job` to the `IQueue`.
+
+4. The `IJobExecutor` will pull the `Job` from the `IQueue` and execute it.
+
+5. The `Job` will be executed by the `IReactor`, which will:
+
+- Create an `Operation` for the `ACTION_CREATE`. This will result in the creation of a new document with a filled out `header` scope but an empty, `{}`, default scope (`global`).
```

Comment:
Should it be the `global` scope or a new `document` scope?
This `document` scope would also have the `SET_DOCUMENT_NAME` and `SET_META` operations.
For the upgrades we could either support a single upgrade operation at the document-level or one per scope.

Reply:
Actually, we already have `header`. I think `SET_DOCUMENT_NAME` and `SET_META` have to affect the `header`. This scope is what is auto-populated and would be the "low-bandwidth" scope that we, for instance, populate a folder structure with.

Reply:
Ah I didn't get to that part yet.
So `header` would also be a scope and be changed with operations, but instead of being inside the `state` object, it is at the top level of the `Document` object?

I like this approach, might be a bit harder to understand for end users but it also means that it will be out of their way if they just care about the state.

Yesterday I started to add the `document` scope and adding it to `state` field was not looking great.